pkg/util/block.go

```diff
@@ -0,0 +1,72 @@
+package util
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	commonTypes "github.com/longhorn/go-common-libs/types"
+	"github.com/longhorn/go-spdk-helper/pkg/types"
+	helperutil "github.com/longhorn/go-spdk-helper/pkg/util"
+)
+
+func GetDevNameFromBDF(bdf string) (string, error) {
+	ne, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create executor")
+	}
+
+	cmdArgs := []string{"-n", "--nodeps", "--output", "NAME"}
+	output, err := ne.Execute(nil, "lsblk", cmdArgs, types.ExecuteTimeout)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to list block devices")
+	}
+
+	devices := strings.Fields(string(output))
+	for _, dev := range devices {
+		link, err := os.Readlink("/sys/block/" + dev)
+		if err != nil {
+			logrus.WithError(err).Warnf("Failed to read link for %s", dev)
+			continue
+		}
+
+		if strings.Contains(link, bdf) {
+			return dev, nil
+		}
+	}
+
+	return "", fmt.Errorf("failed to find device for BDF %s", bdf)
+}
+
+func GetBlockDiskType(devPath string) (string, error) {
+	ne, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create executor")
+	}
+
+	cmdArgs := []string{"-n", "-d", "-o", "type", devPath}
+	output, err := ne.Execute(nil, "lsblk", cmdArgs, types.ExecuteTimeout)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get disk type")
+	}
+
+	return strings.TrimSpace(string(output)), nil
+}
+
+func GetBlockDiskSubsystems(devPath string) ([]string, error) {
+	ne, err := helperutil.NewExecutor(commonTypes.ProcDirectory)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create executor")
+	}
+
+	cmdArgs := []string{"-n", "-d", "-o", "subsystems", devPath}
+	output, err := ne.Execute(nil, "lsblk", cmdArgs, types.ExecuteTimeout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get disk subsystems")
+	}
```

Comment:
We can ask `lsblk` to print out a JSON output with option `-J` then marshal it as a struct. Then:
1. The upper layer does not need to check keywords line by line.
2. We don't need so many kinds of `GetBlockDiskXXXX` functions.

Reply:
It is interesting.
I tried to address the issue, but I found `subsystems` is not printed in the output of `lsblk -n -d -J <path>`.
When I added `lsblk -n -d -J -o subsystems <path>`, the results only contains `subsystems`, like
```
# lsblk -n -d -J -o subsystems /dev/nvme0n1
{
   "blockdevices": [
      {
         "subsystems": "block:nvme:pci"
      }
   ]
}
```
That means we still need to parse the results.

Currently, we only need to get subsystems (GetBlockDiskType is removed. Only GetBlockDiskSubsystem is used). Can we improve it later if we want to get more data?

Reply:
`lsblk -O -J` should print out all fields. You can ignore marshaling unnecessary fields now:
```
lsblk -O -J /dev/sdb
{
   "blockdevices": [
      {
         "name": "sdb",
         "kname": "sdb",
         "path": "/dev/sdb",
         "maj:min": "8:16",
         "fsavail": null,
         "fssize": null,
         "fstype": null,
         "fsused": null,
         "fsuse%": null,
         "fsroots": [
             null
         ],
         "fsver": null,
         "mountpoint": null,
         "mountpoints": [
             null
         ],
         "label": null,
         "uuid": null,
         "ptuuid": null,
         "pttype": null,
         "parttype": null,
         "parttypename": null,
         "partlabel": null,
         "partuuid": null,
         "partflags": null,
         "ra": 128,
         "ro": false,
         "rm": false,
         "hotplug": false,
         "model": "Micron_5200_MTFD",
         "serial": "2005264A2CC6",
         "size": "447.1G",
         "state": "running",
         "owner": "root",
         "group": "disk",
         "mode": "brw-rw----",
         "alignment": 0,
         "min-io": 4096,
         "opt-io": 0,
         "phy-sec": 4096,
         "log-sec": 512,
         "rota": false,
         "sched": "mq-deadline",
         "rq-size": 64,
         "type": "disk",
         "disc-aln": 0,
         "disc-gran": "4K",
         "disc-max": "2G",
         "disc-zero": false,
         "wsame": "0B",
         "wwn": "0x500a0751264a2cc6",
         "rand": false,
         "pkname": null,
         "hctl": "1:0:0:0",
         "tran": "sata",
         "subsystems": "block:scsi:pci",
         "rev": "U020",
         "vendor": "ATA     ",
         "zoned": "none",
         "dax": false
      }
   ]
}
```

Reply:
Addressed. Thank you.